src/state_migration/nv21/miner.rs

```diff
@@ -0,0 +1,293 @@
+                }
+            } else {
+                let in_deadline = store
+                    .get_cbor::<DeadlineOld>(deadline)?
+                    .context("failed to get in_deadline")?;
+
+                let out_sectors_snapshot_cid_cache_key =
+                    sectors_amt_key(&in_deadline.sectors_snapshot)?;
+
+                let out_sectors_snapshot_cid = cache.get_or_insert_with(
+                    out_sectors_snapshot_cid_cache_key,
+                    || -> anyhow::Result<Cid> {
+                        let prev_in_root = cache.get(&miner_prev_sectors_in_key(address));
+                        let prev_out_root = cache.get(&miner_prev_sectors_out_key(address));
+
+                        if let (Some(prev_in_root), Some(prev_out_root)) =
+                            (prev_in_root, prev_out_root)
+                        {
+                            self.migrate_sectors_with_diff(
+                                store,
+                                &in_deadline.sectors_snapshot,
+                                &prev_in_root,
+                                &prev_out_root,
+                            )
+                        } else {
+                            let in_sector_snapshot =
+                                ArrayOld::load(&in_deadline.sectors_snapshot, store)?;
+
+                            let mut out_snapshot =
+                                self.migrate_sectors_from_scratch(store, &in_sector_snapshot)?;
+
+                            Ok(out_snapshot.flush()?)
+                        }
+                    },
+                )?;
+
+                let out_deadline = DeadlineNew {
+                    partitions: in_deadline.partitions,
+                    expirations_epochs: in_deadline.expirations_epochs,
+                    partitions_posted: in_deadline.partitions_posted,
+                    early_terminations: in_deadline.early_terminations,
+                    live_sectors: in_deadline.live_sectors,
+                    total_sectors: in_deadline.total_sectors,
+                    faulty_power: TypeMigrator::migrate_type(in_deadline.faulty_power, store)?,
+                    optimistic_post_submissions: in_deadline.optimistic_post_submissions,
+                    sectors_snapshot: out_sectors_snapshot_cid,
+                    partitions_snapshot: in_deadline.partitions_snapshot,
+                    optimistic_post_submissions_snapshot: in_deadline
+                        .optimistic_post_submissions_snapshot,
+                };
+
+                let out_deadline = store.put_cbor_default(&out_deadline)?;
+                if i < out_deadlines.due.len() {
+                    out_deadlines.due[i] = out_deadline;
+                } else {
+                    out_deadlines.due.push(out_deadline);
+                }
+            }
+        }
+        store.put_cbor_default(&out_deadlines)
+    }
+}
+
+fn sectors_amt_key(sectors: &Cid) -> anyhow::Result<String> {
```

Comment:
The same functions exist in nv17/miner.rs as well, should we reuse them or use the same prefix naming rule? See https://github.com/ChainSafe/forest/pull/2983#discussion_r1244942545

Reply:
I'd say let's postpone refactoring of this (and some other methods, like `migrate_sectors_from_scratch`) for the 3rd time they appear. https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
Does it sound okay for you?

Reply:
Sounds good